src/Network/WebSockets/TSSocketManager.m

```diff
@@ -425,157 +461,234 @@ - (NSString *)webSocketAuthenticationString {
-        return YES;
-    } else if (self.backgroundKeepAliveTimer ||
-               self.backgroundConnectTimer ||
-               self.fetchingTaskIdentifier != UIBackgroundTaskInvalid) {
-        // If app is doing any work in the background, keep web socket alive.
-        return YES;
+#pragma mark UI Delegates
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary *)change
+                       context:(void *)context
+{
+    if (context == kSocketStatusObservationContext) {
+        [self notifyStatusChange];
     } else {
-        return NO;
+        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
     }
 }
 
-- (void)scheduleRetry {
-    OWSAssert([NSThread isMainThread]);
-   
-    if (![self shouldKeepWebSocketAlive]) {
-        // Don't bother retrying if app is inactive and not doing any background activity.
-        [self.reconnectTimer invalidate];
-        self.reconnectTimer = nil;
-    } else if (![self.reconnectTimer isValid]) {
-        // TODO: It'd be nice to do exponential backoff.
-        self.reconnectTimer = [NSTimer timerWithTimeInterval:kWebSocketReconnectTry
-                                                      target:[self class]
-                                                    selector:@selector(becomeActive)
-                                                    userInfo:nil
-                                                     repeats:YES];
-        // Additionally, we want the reconnect timer to work in the background too.
-        [[NSRunLoop mainRunLoop] addTimer:self.reconnectTimer forMode:NSDefaultRunLoopMode];
-    } else {
-        DDLogWarn(@"Not scheduling retry, valid timer");
+- (void)notifyStatusChange
+{
+    switch (self.status) {
+        case kSocketStatusOpen:
+            [[NSNotificationCenter defaultCenter] postNotificationName:SocketOpenedNotification object:self];
+            break;
+        case kSocketStatusClosed:
+            [[NSNotificationCenter defaultCenter] postNotificationName:SocketClosedNotification object:self];
+            break;
+        case kSocketStatusConnecting:
+            [[NSNotificationCenter defaultCenter] postNotificationName:SocketConnectingNotification object:self];
+            break;
+        default:
+            break;
     }
 }
 
-#pragma mark - Background Connect
-
-+ (void)becomeActiveFromForeground {
-    DispatchMainThreadSafe(^{
-        [[self sharedManager] becomeActiveFromForeground];
-    });
++ (void)sendNotification
+{
+    [[self sharedManager] notifyStatusChange];
 }
 
-- (void)becomeActiveFromForeground {
+#pragma mark - Socket LifeCycle
+
+- (BOOL)shouldSocketBeOpen
+{
     OWSAssert([NSThread isMainThread]);
-    
-    if (self.fetchingTaskIdentifier != UIBackgroundTaskInvalid) {
-        [self closeBackgroundTask];
+
+    if (![TSAccountManager isRegistered]) {
```

Comment:
We should avoid introducing any blocking DB i/o to our app launch startup path, since it can cause startup to hang. e.g. put a break point here and see how the app hangs on the launch screen.

I've seen DB reading result in a repeated crash-on-launch when a new DB extension is added. It's tricky because it bites you later. Specifically, adding a new db extension enumerates all objects, which can take a long time if the user has a large message database and is on an old device.

Similarly, it looks like we're also reading from the DB in `ensureWebsocketIsOpen` in order to get the websocket credentials. We previously got the websocket credentials in `ensureWebsocket` but that dispatched async, which maybe sidestepped the issue.

AFAICT this is only the result of the newly observed UIApplicationDidBecomeActiveNotification. Can we delay observing that notification until the app has successfully launched?